src/util/bson/indexed_map.hpp

```diff
@@ -0,0 +1,239 @@
+
+    class iterator
+    {
+        size_t m_idx = 0;
+        IndexedMap* m_map;
+    public:
+
+        iterator(IndexedMap* map, size_t idx);
+        entry operator*()
+        {
+            return m_map->operator[](m_idx);
+        }
+        iterator& operator++();
+        iterator& operator--();
+        iterator operator++(int);
+        iterator operator--(int);
+
+        bool operator!=(const typename IndexedMap<T>::iterator& rhs) noexcept
+        {
+            return !(m_idx == rhs.m_idx);
+        }
+    };
+
+    IndexedMap();
+    IndexedMap(std::initializer_list<entry> entries);
+
+    /// The size of the map
+    size_t size() const;
+
+    /// Find an entry by index
+    entry operator[](size_t idx);
+
+    iterator begin();
+
+    iterator end();
+
+    iterator front();
+
+    iterator back();
+
+    /// Find or add a given key
+    T& operator[](const std::string& k);
+
+    /// Whether or not this map is empty
+    bool empty();
+
+    /// Pop the last entry of the map
+    void pop_back();
+
+    const std::vector<std::string> keys() const;
+private:
+    template <typename V>
+    friend inline bool operator==(const IndexedMap<V>& lhs, const IndexedMap<V>& rhs) noexcept;
+    template <typename V>
+    friend inline bool operator!=(const IndexedMap<V>& lhs, const IndexedMap<V>& rhs) noexcept;
+    std::unordered_map<std::string, T> m_map;
+    std::vector<std::string> m_keys;
+};
+
+template <typename T>
+inline bool operator==(const typename IndexedMap<T>::iterator& lhs,
+                       const typename IndexedMap<T>::iterator& rhs) noexcept
+{
+    return lhs.m_idx == rhs.m_idx;
+}
+
+template <typename T>
+inline bool operator!=(const typename IndexedMap<T>::iterator lhs,
+                       const typename IndexedMap<T>::iterator rhs) noexcept
+{
+    return !(lhs.m_idx == rhs.m_idx);
+}
+
+template <typename T>
+inline bool operator==(const IndexedMap<T>& lhs, const IndexedMap<T>& rhs) noexcept
+{
+    return lhs.m_map == rhs.m_map && lhs.m_keys == rhs.m_keys;
+}
+
+template <typename T>
+inline bool operator!=(const IndexedMap<T>& lhs, const IndexedMap<T>& rhs) noexcept
+{
+    return lhs.m_map != rhs.m_map && lhs.m_keys != rhs.m_keys;
+}
+
+template <typename T>
+IndexedMap<T>::iterator::iterator(IndexedMap<T>* map, size_t idx) : m_idx(idx), m_map(map) {}
+
+template <typename T>
+typename IndexedMap<T>::iterator& IndexedMap<T>::iterator::operator++()
+{
+    m_idx++;
+    return *this;
+}
+
+template <typename T>
+typename IndexedMap<T>::iterator& IndexedMap<T>::iterator::operator--()
+{
+    m_idx--;
+    return *this;
+}
+
+template <typename T>
+typename IndexedMap<T>::iterator IndexedMap<T>::iterator::operator++(int)
+{
+    return IndexedMap<T>::iterator(m_idx++);
+}
+
+template <typename T>
+typename IndexedMap<T>::iterator IndexedMap<T>::iterator::operator--(int)
+{
+    return --(*this);
+}
+
+template <typename T>
+using entry = std::pair<std::string, T>;
```

Comment:
This is declaring a separate template alias for `entry` outside of the scope of `IndexedMap`, and it shouldn't be.